csharp/src/Ice/Connection.cs

```diff
@@ -50,39 +52,20 @@ public ConnectionInfo(ConnectionInfo underlying, bool incoming, string adapterNa
 
     public delegate void CloseCallback(Connection con);
 
-    /// <summary>
-    /// Determines the behavior when manually closing a connection.
-    /// </summary>
-    public enum ConnectionClose
+    public interface Connection
     {
         /// <summary>
-        /// Close the connection immediately without sending a close connection protocol message to the peer and waiting
-        ///  for the peer to acknowledge it.
+        /// Closes the connection forcefully without waiting for any other operation to complete.
         /// </summary>
+        void abort();
 
-        Forcefully,
         /// <summary>
-        /// Close the connection by notifying the peer but do not wait for pending outgoing invocations to complete.
-        /// On the
-        ///  server side, the connection will not be closed until all incoming invocations have completed.
+        /// Closes the connection gracefully.
         /// </summary>
-
-        Gracefully,
-        /// <summary>
-        /// Wait for all pending invocations to complete before closing the connection.
-        /// </summary>
-
-        GracefullyWithWait
-    }
-
-    public interface Connection
-    {
-        /// <summary>
-        /// Manually close the connection using the specified closure mode.
-        /// </summary>
-        /// <param name="mode">Determines how the connection will be closed.
-        ///  </param>
-        void close(ConnectionClose mode);
+        /// <param name="waitForInvocations">If true, this operation waits for all pending invocations to complete
+        /// before closing the connection.</param>
+        /// <returns>A task that completes when the connection is closed.</returns>
```

Comment:
What does it mean for an incoming connection?

Reply:
It's the same for all connections.

You can send outgoing requests on an incoming connection. That's what we call "bidir".

Reply:
I mean, what if I get the connection of an incoming request and maybe then or some point later call closeAsync. Does close take into account ongoing dispatches for that connection?

Reply:
I thought close[Async] already waited for dispatches to complete (once state = StateClosing), but after looking at the code, I don't see how we do it.

In particular:
https://github.com/zeroc-ice/ice/blob/106c9299443791bced745eacfd5d6859505502ad/csharp/src/Ice/ConnectionI.cs#L267

waits for upcallCount (= work items in the thread pool), not dispatches. And it's not clear to me if/where we call this waitUntilFinished when closing a connection.

Maybe @bentoi can shed light on this logic?

Now, in IceRPC (which is maybe what you're thinking about), ShutdownAsync waits for both invocations and dispatches to complete _after_ refusing new invocations:

https://github.com/icerpc/icerpc-csharp/blob/b1f65d8d66b671c9948c7bfb04c3881243ba6a96/src/IceRpc/Internal/IceProtocolConnection.cs#L476
https://github.com/icerpc/icerpc-csharp/blob/b1f65d8d66b671c9948c7bfb04c3881243ba6a96/src/IceRpc/Internal/IceProtocolConnection.cs#L502

It's nicer but not trivial to retrofit in Ice.

Also, with IceRPC, the application cannot grab some connection and shut it down or abort it. You can only shut down a connection indirectly by shutting down its container (ClientConnection, ConnectionCache, Server). Same for abort (DestroyAsync in IceRPC).

Reply:
> And it's not clear to me if/where we call this waitUntilFinished when closing a connection.

It's used to implement `ObjectAdapter::waitForDeactivate` and `Communicator::waitForShutdown`. These methods return once all the connection "activity" is done and this includes application code such as dispatch completion and callbacks (close callback, AMI callbacks, etc).

Reply:
In other words, it's not called when a specific connection is closed manually using `Connection::closeAsync`?